File: lsp/hover.go
package main

import (
	"fmt"
	"strings"

	"github.com/Nv7-Github/bsharp/tokens"
	"github.com/Nv7-Github/bsharp/types"
	"github.com/tliron/glsp"
	protocol "github.com/tliron/glsp/protocol_3_16"
)

func tokID(pos protocol.Position, doc *Document) int {
	if doc.Tokens == nil || doc.IRCache == nil {
		return -1
	}
	id := -1
	for i, tok := range doc.Tokens.Tokens {
		if tok.Pos.Line > int(pos.Line) || (tok.Pos.Line == int(pos.Line) && tok.Pos.Char > int(pos.Character)) {
			break
		}
		id = i
	}
	return id
}

func docHover(context *glsp.Context, params *protocol.HoverParams) (*protocol.Hover, error) {
	// Find token
	doc, exists := Documents[params.TextDocument.URI]
	if !exists {
		return nil, nil
	}
	id := tokID(params.Position, doc)
	if id < 0 {
		return nil, nil
	}

	// Check if function, get name
	var name *string = nil
	prev := doc.Tokens.Tokens[id-1]
	tok := doc.Tokens.Tokens[id]
	if prev.Typ == tokens.TokenTypeLBrack {
		name = &tok.Value
	} else if prev.Typ == tokens.TokenTypeIdent && id > 1 {
		prev2 := doc.Tokens.Tokens[id-2]
		if prev2.Typ == tokens.TokenTypeLBrack {
			name = &tok.Value
		}
	}
	if name == nil {
		return nil, nil
	}

	// Function, check if exists and find pos
	for _, fn := range doc.IRCache.Funcs {
		if fn.Name == *name {
			// Gen text
			text := &strings.Builder{}
			fmt.Fprintf(text, "[FUNC %s", fn.Name)
			for _, par := range fn.Params {
				fmt.Fprintf(text, " [PARAM %s %s]", par.Name, par.Type.String())
			}
			if !types.NULL.Equal(fn.RetType) {
				fmt.Fprintf(text, " [RETURNS %s]", fn.RetType.String())
			}
			text.WriteString("]")

			return &protocol.Hover{
				Contents: protocol.MarkupContent{
					Kind:  protocol.MarkupKindMarkdown,
					Value: fmt.Sprintf("```bsharp\n%s\n```", text.String()),
				},
				Range: &protocol.Range{
					Start: protocol.Position{Line: uint32(tok.Pos.Line), Character: uint32(tok.Pos.Char)},
					End:   protocol.Position{Line: uint32(tok.Pos.EndLine), Character: uint32(tok.Pos.EndChar)},
				},
			}, nil
		}
	}

	return nil, nil
}

File: lsp/src/hover.rs
use tokens::TokenData;

use super::*;

impl Backend {
    pub async fn handle_hover(&self, p: HoverParams) -> Result<Option<Hover>> {
        // Get file
        let state = self.state.lock().await;
        let path = get_path(&p.text_document_position_params.text_document.uri);
        if !state.files.contains_key(&path) {
            return Ok(None);
        }
        let f = state.files.get(&path).unwrap();
        let pos = p.text_document_position_params.position;

        // Get token
        let tok = get_tok(pos, f);
        if tok.is_none() {
            return Ok(None);
        }
        let tok = tok.unwrap();

        // Get hover
        // TODO: Fields
        match &tok.data {
            // Type
            TokenData::TYPE(t) => {
                let val = state.ir.typemap.get(t);
                if let Some(ind) = val {
                    let typ = &state.ir.types[*ind];
                    let hover = Hover {
                        contents: HoverContents::Markup(MarkupContent {
                            kind: MarkupKind::Markdown,
                            value: format!("```bsharp\n{}\n```", typ.typ.data.fmt(&state.ir)),
                        }),
                        range: Some(pos_range(tok.pos)),
                    };
                    return Ok(Some(hover));
                }
                Ok(None)
            }
            TokenData::FUNCTION(f) => {
                let f = state.ir.funcs.iter().find(|x| x.name == *f);
                if f.is_none() {
                    return Ok(None);
                }
                let f = f.unwrap();
                let hover = Hover {
                    contents: HoverContents::Markup(MarkupContent {
                        kind: MarkupKind::Markdown,
                        value: format!("```bsharp\n{}\n```", fn_string(&f, &state.ir)),
                    }),
                    range: Some(pos_range(tok.pos)),
                };
                Ok(Some(hover))
            }
            TokenData::VARIABLE(v) => {
                // Search for var
                let mut var = state.ir.variables.iter().find(|x| {
                    x.name == *v
                        && state.ir.scopes[x.scope].pos.start_line <= tok.pos.start_line
                        && state.ir.scopes[x.scope].pos.end_line >= tok.pos.end_line
                        && path
                            .ends_with(&state.ir.fset.files[state.ir.scopes[x.scope].pos.file].name)
                });
                if var.is_none() {
                    var = state.ir.variables.iter().find(|x| x.name == *v);
                    if var.is_none() {
                        return Ok(None);
                    }
                }
                let var = var.unwrap();

                let hover = Hover {
                    contents: HoverContents::Markup(MarkupContent {
                        kind: MarkupKind::Markdown,
                        value: format!("```bsharp\n{}\n```", var.typ.data.fmt(&state.ir)),
                    }),
                    range: Some(pos_range(tok.pos)),
                };
                Ok(Some(hover))
            }
            _ => Ok(None),
        }
    }
}
